fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/ExplainTokens.java

```diff
@@ -0,0 +1,1110 @@
+    }
+
+    @Nonnull
+    protected int[] getTokenSizes() {
+        return tokenSizes;
+    }
+
+    public int getTokenSize(final int explainLevel) {
+        return explainLevel < tokenSizes.length ? tokenSizes[explainLevel] : 0;
+    }
+
+    @Nonnull
+    protected int[] getMinLengths() {
+        return minLengths;
+    }
+
+    public int getMinLength(final int explainLevel) {
+        return explainLevel < minLengths.length ? minLengths[explainLevel] : 0;
+    }
+
+    @Nonnull
+    protected int[] getMaxLengths() {
+        return maxLengths;
+    }
+
+    public int getMaxLength(final int explainLevel) {
+        return explainLevel < maxLengths.length ? maxLengths[explainLevel] : 0;
+    }
+
+    @Nonnull
+    public CharSequence render(@Nonnull final ExplainFormatter formatter) {
+        return render(ExplainLevel.ALL_DETAILS, formatter, Integer.MAX_VALUE);
+    }
+
+    /**
+     * Render all tokens on the given explain level, using the given formatter, and a given budget of number of
+     * characters. If the budget is limited (if {@code remainingCharacterBudget < Integer.MAX_VALUE}), this method will
+     * at worst render into a string of that budget (plus three characters {@code ...} if the rendered char sequence has
+     * to be cut off).
+     * @param renderingExplainLevel the desired {@link ExplainLevel}
+     * @param formatter the formatter
+     * @param remainingCharacterBudget the budget of the number of remaining characters
+     * @return a {@link CharSequence} representing the rendered explain string which is potentially cut off if
+     *         depending on the budget. The returned character sequence is at most of length
+     *         {@code remainingCharacterBudget + 3} to accommodate for the {@code ...}. Note that using
+     *         {@link CharSequence} instead of {@link String} is preferable as {@link StringBuilder} is a
+     *         {@link CharSequence} allowing us to return un-built string builders without materializing them to
+     *         strings.
+     */
+    @Nonnull
+    public CharSequence render(final int renderingExplainLevel,
+                               @Nonnull final ExplainFormatter formatter,
+                               int remainingCharacterBudget) {
+        final StringBuilder stringBuilder = new StringBuilder();
+        for (int i = 0; i < tokens.size(); i++) {
+            final var token = tokens.get(i);
+            final var stringedToken =
+                    token.render(renderingExplainLevel, formatter, remainingCharacterBudget);
+            stringBuilder.append(stringedToken);
+            remainingCharacterBudget -= (remainingCharacterBudget == Integer.MAX_VALUE ? 0 : stringedToken.length());
+            if (remainingCharacterBudget < 1) {
+                if (remainingCharacterBudget == 0 && i + 1 < tokens.size()) {
+                    // if any of the remaining tokens render to more than an empty string we have to append ... instead
```

Comment:
It's not obvious to me where the `...` comes from, but I take it this is in the rendering of the individual token?